README.md
# Lightning-UI

Lightning-UI provides a set of components that are used very often when you are developing in Lightning, with the following components we hope to help you out with setting up your app a bit faster: 

* [CollectionWrapper](docs/CollectionWrapper/index.md)
    * [List](docs/CollectionWrapper/List.md)
    * [Carousel](docs/CollectionWrapper/Carousel.md)
    * [Grid](docs/CollectionWrapper/Grid.md)
* [Keyboard](docs/Keyboard/index.md)
    * [Key](docs/Keyboard/Key.md)
    * [InputField](docs/Keyboard/InputField.md)
* [ScrollingLabel](docs/ScrollingLabel.md)
* [Progress Bar](ProgressBar.md)
* [Stepper](docs/Stepper/index.md)
    * [ArrowStepper](docs/Stepper/ArrowStepper.md)
    * [CarouselStepper](docs/Stepper/CarouselStepper.md)
    * [ProgressStepper](docs/Stepper/ProgressStepper.md)
* [Color Shift](ColorShift.md)

# Install

```bash
npm i @lightningjs/ui
```

## Contributing

The Lightning-SDK is an open source project. If you want to contribute to it, please consider the following:

- the **main** branch is the latest stable release
- the **dev** branch is used for upcoming releases
- all development should be done in dedicated *topic branches* (from latest `dev`-branch)
- please send in your PR against the `dev`-branch

Before you submit your PR, make sure you install the projects dependencies, as this will activate automatic
linting and code formatting in a Git commit hook.

# Examples
In the following repository you can find code examples on how to use the components

[Ui-component-examples](https://github.com/mlapps/ui-component-examples)



CHANGELOG.md
# Changelog
*11 dec 2023*

## v1.3.19 
- When removing the last item from a CollectionWrapper using the remove method and that last item is out of bounds, no longer throws an error.

*17 nov 2023*
## v1.3.17 / v1.3.18
-This fixes an issue that when you use setIndex when the component is inactive and when _cleanUp from Carousel is called, it clears all the items because the scrollTransition.targetValue is not updated.

*25 sep 2023*

## v1.3.16
- fix immediate transition finish when transition is active but immediate is true.
- fix reposition exception (issue #69)

*11 aug 2023*

## v1.3.15
- fix immediate options flow when forcing transition to finish right away.

*17 jul 2023*

## v1.3.14
- fixes scrolltransition resetting when not running.

*5 jul 2023*

## v1.3.13
- fixes the navigation getting stuck when the onRequestItems callback takes a while to resolve

*3 jul 2023*

## v1.3.12
- Focus fix when removing last element

*jun 2023*

## v1.3.11
- Fix remove function CollectionWrapper

## v1.3.10

*30 may 2023*

- Grid setIndex was `async` and now its  `sync` as it was before `v1.3.8`
- Immediate option on setIndex (and other methods) patch is now `sync` instead of an `async` (with duration: 0)

## v1.3.9

*26 may 2023*

- fixed critical exception error
- add options to more methods

## v1.3.8

*17 may 2023*

- add snapToRow functionality to Keyboard
- optimized navigation function Keyboard
- improved Request feature CollectionWrapper
- Resolved minor bugs CollectionWrapper

## v1.3.7

*3 feb 2023*

- Improved Requests/Paging functionality

## v1.3.6

*11 jan 2023*

- Fix Carousel initial scroll
- Fix Carousel infinite loop issue when dataset too small
- Fix Grid indexchanged

## v1.3.5

*3 jan 2023*

- Fix Carousel index issue introduced in v1.3.3. Refactored the code to correctly use internal index to maintain the correct focus and output the data index instead of the internal index.

## v1.3.6
*11 jan 2023*
- Fix Carousel initial scroll
- Fix Carousel infinite loop issue when dataset too small
- Fix Grid indexchanged

## v1.3.5
*3 jan 2023*
- Fix Carousel index issue introduced in v1.3.3. Refactored the code to correctly use internal index to maintain the correct focus and output the data index instead of the internal index.

## v1.3.4

- Fix initial scroll when scroll offset different than 0.
- Fix progress value in ProgressStepper component.

## v1.3.3

- CollectionWrapper now only fires onIndexChanged when the index has actually changed.
- Fix bug where CollectionWrapper could not remove an item.
- Fix Carousel where index does not represent actual index of the focused item.

## v1.3.2

- Fixed name error where code used marginUp instead of marginTop as indicated in documentation

## v1.3.1

- Bug fixed where the Grid would navigation would go to the wrong index when you navigate after useing `setIndex`
- Jest support added for development

## v1.3.0

*22 aug 2022*

- Added Stepper component
- Added ProgressBar component
- Added ColorShift component for color accessibility

## v1.2.8

*9 aug 2022*

- Fixed bug where transition is not reset when the CollectionWrapper is cleared.

## v1.2.7

*18 jul 2022*

- Fixed bug where the width is not calculated when arrow functions (for example; w => w) are used in the template

## v1.2.6

*13 may 2022*

- The Keyboard component now sends more data when firing signals.

## v1.2.5

*25 apr 2022*

- Fixed bug where Carousel cleans up too many ItemWrappers

## v1.2.4

*11 apr 2022*

- Fixed refactor bug in ScrollingLabel
- Fixed autoStart setter / getter not working in ScrollingLabel

## v1.2.3

*8 apr 2022*

- Fixed bug where ItemWrappers are misplaced when navigating too fast in Carousel

## v1.2.2

*30 mar 2022*

- Fixed bug where previous ItemWrapper data is not cleared when the Grid is cleared

## v1.2.1

*21 mar 2022*

- Added wrap around feature to Keyboard
- Fixed bug where start position is not correct in the InputField

## v1.2.0

*25 nov 2021*

- Added a ScrollingLabel component
- Fixed bug in InputField component where cursor is not in the right position

## v1.1.7

*23 nov 2021*

- Added forceLoad feature to the CollectionWrapper

## v1.1.6

*10 nov 2021*

- Fixed bug where maxCharacters are not checked properly in Keyboard
- Fixed bug in navigation calculation resulting in incorrect focus
- Keyboard now resets focus when switching input components
- Added password mode to InputField
- Added auto hide cursore feature to InputField

## v1.1.5

*25 oct 2021*

- Fixed bug where scroll value is incorrect when CollectionWrapper is inactive.

## v1.1.4

*13 oct 2021*

- Fixed parameters for previous navigation Keyboard component.

## v1.1.3

*13 oct 2021*

- Fixed bug where KeyWrapper does not update or respawns the key properly.
- Fixed navigation over rows Keyboard component
- Fixed onInputChanged signal firing when max characters reached Keyboard component
- Added getter and setter for the label tag in Key component

## v1.1.0

*11 oct 2021*

- Released Keyboard component
- Released Key component
- Released InputField component
- Added related docs

## v1.0.10

*24 sep 2021*

- Fixed bug where Grid sets index when target index is the same as current index

## v1.0.9

*15 sep 2021*

- Fixed scroll transition value when CollectionWrapper is inactive
- Fixed navigation issue when navigating up or down

## v1.0.8

*12 jul 2021*

- Fixed bug where items are not position correctly when reposition function is called
- Added margin data to ItemWrapper

## v1.0.7

*5 jul 2021*

- Added feature where the CollectionWrapper returns the component instead of data when the related component is spawned.
- Fixed refactor issue in CollectionWrapper

## v1.0.5

*30 jun 2021*

- Fixed issue where Carousel required a margin to position its items correctly
- Fixed scroll issue in CollectionWrapper

## v1.0.3

*22 jun 2021*

- Fixed setIndex function in Carousel

## v1.0.2

*10 jun 2021*

- Fixed scroll and plot functions Carousel









src/Stepper.ts
import Lightning from '@lightningjs/core';
import type { InlineElement } from '@lightningjs/core/dist/src/tree/Element.d.mts';

export interface StepperTemplateSpec extends Lightning.Component.TemplateSpec {
    focusColor?: number,
    labelColor?: number,
    labelColorFocused?: number,
    padding?: number,
    max?: number,
    min?: number,
    value?: number,
    options?: (string | StepperOption)[],
    focusAnimation?: Lightning.types.Animation | null,

    Focus: object,
    Label: object,
    ValueWrapper: {
        Value: object
    }
}

interface StepperOption {
    label: string
}

export interface StepperEvent {
    value: number,
    options?: StepperOption[]
}

export interface StepperSignalMap extends Lightning.Component.SignalMap {
    onValueChanged(event: StepperEvent): void
}


export interface StepperTypeConfig extends Lightning.Component.TypeConfig {
    SignalMapType: StepperSignalMap
}


export default class Stepper<
    TemplateSpec extends StepperTemplateSpec = StepperTemplateSpec,
    TypeConfig extends StepperTypeConfig = StepperTypeConfig
>
    extends Lightning.Component<TemplateSpec, TypeConfig>
    implements Lightning.Component.ImplementTemplateSpec<StepperTemplateSpec>
{
    Focus: Lightning.Element = (this as Stepper).getByRef('Focus')!;
    Label = (this as Stepper).getByRef('Label')!;
    /**
     * @privateRemarks
     * We explicitly allow ValueWrapper to be extended by subclasses, so we need to make this intention known
     * to the TypeScript compiler so that when subclasses access ValueWrapper, they can access the extended
     * version of ValueWrapper.
     */
    ValueWrapper = (this as Stepper).getByRef('ValueWrapper')! as Lightning.Element<InlineElement<TemplateSpec['ValueWrapper']>>;
    /**
     * @privateRemarks
     * Normally we wouldn't have to explicity assert the type of Value in this case, but there seems to be a bug in the
     * TypeScript compiler where if we don't do that here it causes VSCode to show inconsistent errors throughout the
     * project.
     */
    Value = (this as Stepper).ValueWrapper.getByRef('Value')! as Lightning.Element;

    protected _focusColor = 0xff009245;
    protected _labelColor = 0xff9d9d9d;
    protected _labelColorFocused = 0xffffffff;
    protected _padding = 30;
    protected _max = 100;
    protected _min = 0;
    protected _value = 50;
    protected _options: StepperOption[] = [];
    protected _label = "label";
    protected _focusAnimation: Lightning.types.Animation | null = null;

    static override _template(): Lightning.Component.Template<StepperTemplateSpec> {
        return {
            h: 80, w: 574,
            Focus: {alpha: 0, w: w => w, h: h => h, rect: true},
            Label: {x: 30, y: h => h * 0.5, mountY: 0.5, text: {text: '', fontSize: 22}},
            ValueWrapper: {
                x: w => w - 30, w: 200, h: h => h, mountX: 1,
                Value: {x: w => w * 0.5, y: h => h * 0.5, mountX: 0.5, mountY: 0.5, text: {text: '', fontSize: 22}},
            }
        }
    }

    override _firstActive() {
        if(!this._focusAnimation) {
            this._createFocusAnimation();
        }
    }

    override _focus() {
        if(this._focusAnimation) {
            this._focusAnimation.start();
        }
    }

    override _unfocus() {
        if(this._focusAnimation) {
            this._focusAnimation.stop();
        }
    }

    override _handleLeft() {
        this.navigate(-1);
    }

    override _handleRight() {
        this.navigate(1);
    }

    protected _createFocusAnimation() {
        this._focusAnimation = (this as Stepper).animation({duration: 0.2, stopMethod: 'reverse', actions: [
            {t: 'Focus', p: 'alpha', v: {0: 0, 1: 1}},
            {t: 'Label', p: 'color', v: {0: this._labelColor, 1: this._labelColorFocused}},
            {t: 'ValueWrapper.Value', p: 'color', v: {0: this._labelColor, 1: this._labelColorFocused}},
        ]});
    }

    private navigate(dir: number) {
        this.value = this.calcCarouselNavigation(dir, this._value, this._min, this._max);
        const event: StepperEvent = {
            value: this._value,
        }
        if(this._options) {
            event.options = this._options
        }
        this.fireAncestors('$onValueChanged', event);
        (this as Stepper).signal('onValueChanged', event);
    }

    private calcCarouselNavigation(dir: number, current: number, min: number, max: number) {
        let target = current + dir;
        if(target < min) {
            target = max;
        }
        if(target > max) {
            target = min;
        }
        return target;
    }

    _update() {
        (this as Stepper).patch({
            Focus: {color: this._focusColor},
            Label: {x: this._padding, color: this._labelColor, text: {text: this._label}},
            ValueWrapper: {x: w => w - this._padding,
                Value: {color: this._labelColor, text: {['text' as string]: this.optionValue || this.value}}
            }
        })
        if(this.hasFocus()) {
            this._focus();
        }
    }

    set label(v: string) {
        this._label = v;
        if(this.active) {
            this.Label.text!.text = v;
        }
    }

    get label() {
        return this._label;
    }

    set value(v: number) {
        this._value = v;
        if(this.active) {
            this.Value.text!.text = v.toString();
        }
    }

    get value() {
        return this._value;
    }

    get optionValue() {
        const option = this._options[this._value];
        return option && option.label || undefined;
    }

    set options(v: Array<StepperOption | string>) {
        const refactor = v.map((option) => {
            if(typeof option === 'string') {
                return {label: option};
            }
            return option;
        });

        this._value = 0;
        this._options = refactor;
        this._max = refactor.length - 1;
        this._update();
    }

    get options(): StepperOption[] {
        return this._options;
    }

    set focusColor(v: number) {
        this._focusColor = v;
    }

    get focusColor() {
        return this._focusColor;
    }

    set labelColor(v: number) {
        this._labelColor = v;
    }

    get labelColor() {
        return this._labelColor;
    }

    set labelColorFocused(v: number) {
        this._labelColorFocused = v;
    }

    get labelColorFocused() {
        return this._labelColorFocused;
    }

    set padding(v: number) {
        this._padding = v;
    }

    get padding() {
        return this.padding;
    }

    set max(v: number) {
        this._max = v;
    }

    get max() {
        return this._max;
    }

    set min(v: number) {
        this._min = v;
    }

    get min() {
        return this._min;
    }

    set focusAnimation(v: Lightning.types.Animation | null) {
        this._focusAnimation = v;
    }

    get focusAnimation() {
        return this._focusAnimation
    }
}